### ISSUE TITLE
JmDNS is not working in Android Oreo 8.x

### ISSUE THREAD

**Issue by teocci** [jmdns/jmdns#168]
I have a working on an app which register custom services using JmDNS. It seems that the library stopped working on devices with Android 8.x.

My constructor is:
```
try {
    jmDNS = JmDNS.create(obtainIPv4Address(info));
} catch (IOException e) {
    LogHelper.e(TAG, "Error in JmDNS creation: " + e);
}
```

And my method to register the service is:
```
public void registerService()
{
    Log.e(TAG, "registerService");

    unregisterService();

    final String serviceName = getDeviceID(context.getContentResolver());
    final ServiceInfo serviceInfo = ServiceInfo.create(SERVICE_TYPE, serviceName, DEFAULT_RTSP_PORT, stationName);

    new JmDNSTask(jmDNS, serviceInfo, this).execute();
}
```

Are there any known issues or even known fixes? I didn't get any error.

**Comment by verma171** [comment]
@teocci  I am facing the same issue. Did you find any thing ?

**Comment by evbarnett** [comment]
Are you [obtaining a multicast lock?](https://stackoverflow.com/questions/23805893/android-jmdns-doesnt-discover-devices/23854825#23854825)

Also- what is JmDNSTask doing? What is the value of SERVICE_TYPE? Posting more information would help.

**Comment by verma171** [comment]
@evbarnett  Yes, I am using multicast lock , below is code which I am using 
`public class JmDNSDiscovery extends Thread{
    public static final String SERVICE_TYPE = "_http._tcp.local.";

    private JmDNS _jmdns;
    private android.net.wifi.WifiManager.MulticastLock  _multicastLock;
    private Context context;
    private ServiceListener _serviceListener;

    public JmDNSDiscovery(Context context){
        this.context = context;
    }

    public void run(){

        // Enable device to multicast
        // see: http://developer.android.com/reference/android/net/wifi/WifiManager.MulticastLock.html
//        _multicastLock = ((WifiManager)context.getSystemService(WIFI_SERVICE)).createMulticastLock("MultiCastLock");
//        _multicastLock.setReferenceCounted(true);
//        _multicastLock.acquire();

        android.net.wifi.WifiManager wifi =
                (android.net.wifi.WifiManager)
                        context.getSystemService(android.content.Context.WIFI_SERVICE);
        _multicastLock = wifi.createMulticastLock("HeeereDnssdLock");
        _multicastLock.setReferenceCounted(true);
        _multicastLock.acquire();

        // Setup JmDNS
        try {

           _jmdns = JmDNS.create(InetAddress.getLocalHost());
         //   _jmdns = JmDNS.create();
            _serviceListener = new ServiceListener() {
                public void serviceAdded(ServiceEvent event) {
                    Log.d("Service added: " , String.valueOf(event.getInfo()));
                }
                public void serviceRemoved(ServiceEvent event) {
                    Log.d("Service removed: " , String.valueOf(event.getInfo()));
                }
                public void serviceResolved(ServiceEvent event) {
                    Log.d("Service resolved: " , String.valueOf(event.getInfo()));
                    int port = event.getInfo().getPort();
                    InetAddress host = event.getInfo().getInetAddresses()[0];
                    SendMessage sendMessageTask = new SendMessage(host, port);
                    sendMessageTask.execute();
                }
            };

            _jmdns.addServiceListener(SERVICE_TYPE, _serviceListener);

            Thread.sleep(2*60*1000L);

            if (_multicastLock != null) _multicastLock.release();

            System.out.println("Discovery over");

        } catch (Exception e) {
            // TODO write exception code
            e.printStackTrace();
        }
    }

    private class SendMessage extends AsyncTask<Void, Void, Void> {
        int port;
        InetAddress host;

        public SendMessage(InetAddress host, int port) {
            this.host = host;
            this.port = port;
        }

        @Override
        protected Void doInBackground(Void... params) {

            try (Socket clientSocket = new Socket(host, port);){
                DataOutputStream outToServer =
                        new DataOutputStream(clientSocket.getOutputStream());
                outToServer.writeBytes("Hi from client "+new Date()+"\n");
                InputStream inputStream = clientSocket.getInputStream();
                BufferedReader inFromServer = new BufferedReader(new InputStreamReader(inputStream));
                System.out.println("Response from Discovered Server: " + inFromServer.readLine());
            } catch (Exception e) {
                e.printStackTrace();
            }
            return null;
        }
    }

}`

**Comment by evbarnett** [comment]
Make sure you have the right permissions, I think you'll need

```
<uses-permission android:name="android.permission.CHANGE_WIFI_MULTICAST_STATE" />
<uses-permission android:name="android.permission.INTERNET"/>
<uses-permission android:name="android.permission.ACCESS_NETWORK_STATE"/>
<uses-permission android:name="android.permission.ACCESS_WIFI_STATE"/>
```

but I'm not sure.

**Comment by teocci** [comment]
> @teocci I am facing the same issue. Did you find any thing ?

I fixed this by using this constructor:
```
try {
    jmDNS = JmDNS.create(InetAddress.getByName(obtainIPv4Address(info)), HOST_NAME);
} catch (IOException e) {
    LogHelper.e(TAG, "Error in JmDNS creation: " + e);
}
```

**Comment by dtboy1995** [comment]
can you provide your code,thanks a lot

**Comment by melloware** [comment]
The code is all in my StackOverflow post: https://stackoverflow.com/questions/23805893/android-jmdns-doesnt-discover-devices/23854825#23854825

**Comment by chjan** [comment]
Android8 problem: 

I have the code according to melloware, and it is working on A7, but not on my new LG V30 A8. I have verified that the wifi-local ip address is correct, and I do use the recommended method call jmdns.create (adr, "myhostName");

I am stuck, anyone with a remedy?
